runtime: rewrite channel implementation

aykevl opened the pull request:
This rewrite simplifies the channel implementation considerably, with 34% fewer LOC. Perhaps the most important change is the removal of the channel state, which made sense when we had only send and receive operations but only makes things more compliated when multiple select operations can be pending on a single channel.

I did this rewrite originally to make it possible to make channels parallelism-safe (see https://github.com/tinygo-org/tinygo/issues/2446). The current implementation is not parallelism-safe, but it will be easy to make it so (the main additions will be a channel lock, a global select lock, and an atomic compare-and-swap in chanQueue.pop).

This change also reduces binary size: for the driver examples that do change (presumably the ones that use a channel), the code size is reduced by around 400 bytes.

aykevl committed:
runtime: rewrite channel implementation

This rewrite simplifies the channel implementation considerably, with
34% less LOC. Perhaps the most important change is the removal of the
channel state, which made sense when we had only send and receive
operations but only makes things more compliated when multiple select
operations can be pending on a single channel.

I did this rewrite originally to make it possible to make channels
parallelism-safe. The current implementation is not parallelism-safe,
but it will be easy to make it so (the main additions will be a channel
lock, a global select lock, and an atomic compare-and-swap in
chanQueue.pop).

eliasnaur commented on src/runtime/chan.go:
```diff
@@ -1,520 +1,329 @@
-	// Before resuming the task, all other channel operations on this select statement should be canceled by removing them from their corresponding lists.
-	allSelectOps []channelBlockedList
+const (
+	chanOperationWaiting = iota // waiting for a send/receive operation to continue
+	chanOperationOk             // successfully sent or received (not closed)
+	chanOperationClosed         // channel was closed, the value has been zeroed
```
Nit: because of `chanOperationMask`, these would be clearer (to me) spelled out in binary literal form.

aykevl replied:
Seems reasonable, thanks!

eliasnaur commented on src/runtime/chan.go:
```diff
@@ -1,520 +1,329 @@
+// Pop the next waiting channel from the queue. Channels that are no longer
+// waiting (for example, when they're part of a select operation) will be
+// skipped.
+// This function must be called with interrupts disabled.
```
Is 'interrupts disabled' still accurate when our lord and saviour, multithreading support, arrives? Here and below.

aykevl replied:
Yes, though the comment will need to be updated to say something like "this function must be called with the channel lock held and interrupts disabled".  Interrupts still need to be disabled on microcontrollers. What will happen exactly on a chip like the rp2040 (that has both interrupts and two cores) is something I'll figure out when that happens.

eliasnaur commented on src/runtime/chan.go:
```diff
@@ -1,520 +1,329 @@
-	ch.blocked = blockedlist
-	chanDebug(ch)
-	interrupt.Restore(i)
+	// We can't proceed, so we add ourselfs to the list of receivers and wait
```
Nit: "ourselves".

aykevl replied:
Thanks, fixed!

eliasnaur reviewed (commented):
Drive-by review.

deadprogram commented:
@aykevl any responses to @eliasnaur comments? :smile_cat:

deadprogram commented:
Also note that I changed the branch for the PR just now to `dev`.

aykevl commented:
Rebased and fixed the issues pointed out by @eliasnaur.

Note that this PR and #4549 do conflict with each other (even though that won't be visible in git/GitHub). Once one is merged, the other will need to be rebased and updated.

deadprogram commented:
This PR can now be rebased @aykevl if you please.

aykevl commented:
Rebased after merging #4549!

deadprogram reviewed (approved):
Good to reduce complexity, thanks for working on this!

deadprogram commented:
Now merging, thank you to @aykevl for working on this and to @eliasnaur for review.
